src/modules/src/error_estimator_ukf.c

```diff
@@ -0,0 +1,1623 @@
+{
+  quatRes[0] = q1[0] * q2[0] - q1[1] * q2[1] - q1[2] * q2[2] - q1[3] * q2[3];
+  quatRes[1] = q1[1] * q2[0] + q1[0] * q2[1] - q1[3] * q2[2] + q1[2] * q2[3];
+  quatRes[2] = q1[2] * q2[0] + q1[3] * q2[1] + q1[0] * q2[2] - q1[1] * q2[3];
+  quatRes[3] = q1[3] * q2[0] - q1[2] * q2[1] + q1[1] * q2[2] + q1[0] * q2[3];
+}
+
+// compute quaternion from attitude error
+static void quatFromAtt(float *attVec, float *quat)
+{
+  float absError = sqrtf(attVec[0] * attVec[0] + attVec[1] * attVec[1] + attVec[2] * attVec[2]);
+  float absError2 = absError * absError;
+  float absError3 = absError2 * absError;
+  float absError4 = absError3 * absError;
+
+  float scale = 0.5f - 0.02083333f * absError2 + 0.0002604166f * absError4;
+  quat[0] = 1.0f - 0.125f * absError2 + 0.002604166f * absError4;
+
+  quat[1] = scale * attVec[0];
+  quat[2] = scale * attVec[1];
+  quat[3] = scale * attVec[2];
+}
+
+// Temporary development groups
+LOG_GROUP_START(kalman_states)
+//  LOG_ADD(LOG_FLOAT, ox, &coreData.S[KC_STATE_X])
+//  LOG_ADD(LOG_FLOAT, oy, &coreData.S[KC_STATE_Y])
+//  LOG_ADD(LOG_FLOAT, vx, &coreData.S[KC_STATE_PX])
+//  LOG_ADD(LOG_FLOAT, vy, &coreData.S[KC_STATE_PY])
+LOG_GROUP_STOP(kalman_states)
+
+LOG_GROUP_START(navFilter)
+LOG_ADD(LOG_FLOAT, posX, &stateNav[0])
+LOG_ADD(LOG_FLOAT, posY, &stateNav[1])
+LOG_ADD(LOG_FLOAT, posZ, &stateNav[2])
+LOG_ADD(LOG_FLOAT, accX, &accLog[0])
+LOG_ADD(LOG_FLOAT, accY, &accLog[1])
+LOG_ADD(LOG_FLOAT, accZ, &accLog[2])
+LOG_ADD(LOG_FLOAT, omegaX, &omega[0])
+LOG_ADD(LOG_FLOAT, omegaY, &omega[1])
+LOG_ADD(LOG_FLOAT, omegaZ, &omega[2])
+LOG_ADD(LOG_FLOAT, Phi, &eulerOut[0])
+LOG_ADD(LOG_FLOAT, Theta, &eulerOut[1])
+LOG_ADD(LOG_FLOAT, Psi, &eulerOut[2])
+LOG_ADD(LOG_FLOAT, Px, &PxOut)
+LOG_ADD(LOG_FLOAT, Pvx, &PvxOut)
+LOG_ADD(LOG_FLOAT, Pattx, &PattxOut)
+LOG_ADD(LOG_UINT32, nanCounter, &nanCounterFilter)
+LOG_ADD(LOG_FLOAT, range, &rangeCF)
+LOG_ADD(LOG_FLOAT, procTimeFilter, &procTime)
+LOG_ADD(LOG_UINT8, recAnchorId, &receivedAnchor)
+LOG_GROUP_STOP(navFilter)
+
+LOG_GROUP_START(sensorFilter)
+LOG_ADD(LOG_FLOAT, dxPx, &meas_NX)
+LOG_ADD(LOG_FLOAT, dyPx, &meas_NY)
+LOG_ADD(LOG_FLOAT, dxPxPred, &pred_NX)
+LOG_ADD(LOG_FLOAT, dyPxPred, &pred_NY)
+LOG_ADD(LOG_FLOAT, distPred, &predDist)
+LOG_ADD(LOG_FLOAT, distMeas, &measDist)
+LOG_ADD(LOG_FLOAT, baroHeight, &baroOut)
+LOG_ADD(LOG_FLOAT, innoChFlow_x, &innoCheckFlow_x)
+LOG_ADD(LOG_FLOAT, innoChFlow_y, &innoCheckFlow_y)
+LOG_ADD(LOG_FLOAT, innoChTof, &innoCheckTof)
+LOG_ADD(LOG_FLOAT, distTWR, &distanceTWR)
+LOG_GROUP_STOP(sensorFilter)
+
+// Stock log groups
+LOG_GROUP_START(ukf)
+STATS_CNT_RATE_LOG_ADD(rtUpdate, &updateCounter)
+STATS_CNT_RATE_LOG_ADD(rtPred, &predictionCounter)
+STATS_CNT_RATE_LOG_ADD(rtBaro, &baroUpdateCounter)
+STATS_CNT_RATE_LOG_ADD(rtFinal, &finalizeCounter)
+STATS_CNT_RATE_LOG_ADD(rtApnd, &measurementAppendedCounter)
+STATS_CNT_RATE_LOG_ADD(rtRej, &measurementNotAppendedCounter)
+LOG_GROUP_STOP(ukf)
+
+PARAM_GROUP_START(ukf)
```

Comment:
The same can be done for params: https://github.com/bitcraze/crazyflie-firmware/blob/af96c070d6bc32cf73eeb89980a970d7e01f7f8c/src/modules/src/estimator_kalman.c#L588-L595

https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/api/params/#kalman

Reply:
This is optional